src/storage/src/storage_state/async_storage_worker.rs

```diff
@@ -162,54 +196,69 @@ impl<T: Timestamp + Lattice + Codec64 + Display> AsyncStorageWorker<T> {
                         ingestion_description,
                         _phantom_data,
                     ) => {
-                        let mut state = ingestion_description
-                            .initialize_state(&persist_clients)
+                        let mut calc = ingestion_description
+                            .create_calc(&persist_clients)
                             .instrument(span.clone())
                             .await;
-                        let resume_upper: Antichain<T> = ingestion_description
-                            .calculate_resumption_frontier(&mut state)
-                            .instrument(span)
-                            .await;
+
+                        let resume_upper =
+                            calc.calculate_resumption_frontier().instrument(span).await;
+
+                        let export_uppers = calc.get_uppers();
+
+                        // Convenience macro to convert `BTreeMap<GlobalId, Antichain<C>>` to
+                        // `BTreeMap<GlobalId, Vec<Row>>`.
+                        macro_rules! to_vec_row {
```

Comment:
Macros have the highest of bars to justify their existence. If you want to factor this out creating a normal function with a nice doc string is the way to go. Something like:

```rust
/// some doc string
fn encode_frontier<T: SourceTimestamp>(frontier: &Antichain<T>) -> Vec<Row> {
    frontier.iter().map(|ts| ts.encode_row()).collect()
}

// at call site
uppers.iter().map(|(id, upper)| (*id, encode_frontier(upper)).collect()

```

Reply:
I think we feel differently about one-liners! I think introducing 4 generics is less tasteful, but certainly glad to use a function.